### DISCUSSION TITLE
NLP model construction performance

### DISCUSSION THREAD

**Discussion by chmeckel** [casadi/casadi#3283] [Q&A]
Hi, I'm currently dealing with non-linear programming and I have found IPOPT to be quite fast in comparison to another solvers I've tried.
However, CasADi/IPOPT still takes a really long time when executing my MATLAB code (currently about 400 s in total while the actual nlp solver only takes about 2 seconds to find an optimal solution).
When building the model constraints in the MATLAB code, I basically have to create functions that consist of SX/MX symbols, correct? This is the part that seems to be taking quite a long time and I don't know why it does that. Am I using CasADi incorrectly?

For example, this is a rather small portion of my code that takes about 12 s to run once. I'll have to run it 30 times, resulting in 360 s total, meaning that my code only takes 10 % of the time when skipping these few lines.
Does anyone have an idea on how to improve this code? 
```
for k = (1:22)
    Qc = 0;
    for l = (1:11)
        for m = (1:11)
            Qc = Qc+xvar((p-1)*nv+l)*xvar((p-1)*nv+m)*(GKT(k, l)*GKT(k, m)+...
                BKT(k, l)*BKT(k, m))+xvar((p-1)*nv+nb+l)*xvar((p-1)*nv+nb+m)*(GKT(k, l)*GKT(k, m)+...
                BKT(k, l)*BKT(k, m))+xvar((p-1)*nv+nb+l)*xvar((p-1)*nv+m)*(GKT(k, l)*BKT(k, m)-...
                GKT(k, m)*BKT(k, l))+xvar((p-1)*nv+l)*xvar((p-1)*nv+nb+m)*-(GKT(k, l)*BKT(k, m)-...
                GKT(k, m)*BKT(k, l));
        end
    end

    constr = vertcat(constr, Qc);
    lbg(qccnt,:) = 0;
    ubg(qccnt,:) = imax(k)^2;

    qccnt = qccnt+1;
end
```
The `xvar` vector consists of the model variablkes (SX symbols); `GKT` and `BKT` are matrices with constant real elements. All this code basically does is form bilinear combinations of the model variables. The `vertcat` portion also doesn't seem to slow down my code even though MATLAB gives a warning in the code editor.

Thanks in advance!

**Comment by brunoml5** [comment]
Hi!

Take a look at the [for-loop equivalents](https://web.casadi.org/docs/#for-loop-equivalents) section in the documentation. It is much more efficient in memory because it avoids repeting the expressions. 

Also, take a look at this [presentation](https://www.youtube.com/watch?v=Xfcp5RK7XZQ) from Joris Gillis. He explains in details what is happening under the hood in casadi. The part about maps starts about 23:00. With imagination, it is possible to construct very efficient graphs in casadi.

I hope it helps!

**Reply by chmeckel** [reply]
Thank you, this looks like it should help. I don't really understand how to translate my for-loop into one of the equivalents though and I can't find any good examples either. Guess I'll figure it out somehow.

**Reply by brunoml5** [reply]
Hi! Keep in mind "functional programming" while using Casadi.

If you have some expression which is a little bit complicated and repeat many times in your model, encapsulate it in a Casadi Function. The creation of the model will be faster and it will use less memory since there will be only one declaration of the expression.

You can generate C code for your models and take a look at it. It is basically the internal memory representation of the model. You will see that it separates each Casadi Function in a C function.

**Reply by chmeckel** [reply]
Hi, thank you again.
using a Casadi Function has reduced the time spent by a little more than 50 %. My code now looks pretty much like this:
```
GKTL = SX.sym('GKTL');
GKTM = SX.sym('GKTM');
BKTL = SX.sym('GKTL');
BKTM = SX.sym('GKTM');
xvarL = SX.sym('xvarL');
xvarM = SX.sym('xvarM');
xvarnbL = SX.sym('xvarnbL');
xvarnbM = SX.sym('xvarnbM');
e = xvarL*xvarM*(GKTL*GKTM+...
    BKTL*BKTM)+xvarnbL*xvarnbM*(GKTL*GKTM+...
    BKTL*BKTM)+xvarnbL*xvarM*(GKTL*BKTM-...
    GKTM*BKTL)+xvarL*xvarnbM*-(GKTL*BKTM-...
    GKTM*BKTL);
f = Function('f', {GKTL,GKTM,BKTL,BKTM,xvarL,xvarM,xvarnbL,xvarnbM}, {e});

for k = (1:2*nl)
    Qc = 0;
    for l = (1:nb)
        for m = (1:nb)
            Qc = Qc+f(GKT(k, l), GKT(k, m), BKT(k, l), BKT(k, m), xvar((p-1)*nv+l), xvar((p-1)*nv+m), xvar((p-1)*nv+nb+l), xvar((p-1)*nv+nb+m));
        end
    end

    constr = vertcat(constr, Qc);
    lbg(qccnt,:) = 0;
    ubg(qccnt,:) = imax(k)^2;

    qccnt = qccnt+1;
end
```
Now, if I wanted to reduce the time even further, I would have to get rid of the three for-loops. But I don't quite understand how I would use the `map` (or `mapaccum`?) method to iterate through the matrices.

**Reply by brunoml5** [reply]
Hi!

You are writing your problem in a really confusing way. For instance, your Function is just simple arithmetic operations:
```
aux1 = GKTL*GKTM + BKTL*BKTM;
aux2 = GKTL*BKTM - GKTM*BKTL;

e = (xvarL*xvarM + xvarnbL*xvarnbM)*aux1 + (xvarnbL*xvarM- xvarL*xvarnbM)*aux2;

f = Function('f', {GKTL,GKTM,BKTL,BKTM,xvarL,xvarM,xvarnbL,xvarnbM}, {e});
```
Because of that, I've got the feeling that you should define your whole problem using matrices. 

Organize your code better and you might see opportunities to use `map` or `mapaccum`.
